File: src/backends/versitygw.py
import logging
import boto3
import xml.etree.ElementTree as ET

from .backend import Backend
from ..utils import sign_v4_request, send_signed_request

logger = logging.getLogger("s3-resource-operator.versitygw")

class VersityGW(Backend):
    """Backend implementation for VersityGW."""

    def __init__(self, endpoint_url, access_key, secret_key):
        self.endpoint_url = endpoint_url
        self.access_key = access_key
        self.secret_key = secret_key
        self.s3 = boto3.client(
            "s3",
            endpoint_url=endpoint_url,
            aws_access_key_id=access_key,
            aws_secret_access_key=secret_key,
        )

    def create_bucket(self, bucket_name, owner=None):
        if self.bucket_exists(bucket_name):
            logger.info(f"S3 bucket '{bucket_name}' already exists.")
            return
        try:
            self.s3.create_bucket(Bucket=bucket_name)
            logger.info(f"S3 bucket '{bucket_name}' created.")
            if owner:
                self.change_bucket_owner(bucket_name, owner)
        except self.s3.exceptions.BucketAlreadyOwnedByYou:
            logger.info(f"S3 bucket '{bucket_name}' already exists and is owned by you.")
        except self.s3.exceptions.BucketAlreadyExists:
            logger.warning(f"S3 bucket '{bucket_name}' already exists and is owned by someone else.")
        except Exception as e:
            logger.error(f"Error creating bucket '{bucket_name}': {e}")

    def delete_bucket(self, bucket_name):
        try:
            self.s3.delete_bucket(Bucket=bucket_name)
            logger.info(f"S3 bucket '{bucket_name}' deleted.")
        except Exception as e:
            logger.error(f"Error deleting bucket '{bucket_name}': {e}")

    def _list_buckets_raw(self):
        signed_request = sign_v4_request(
            method="PATCH",
            url=f"{self.endpoint_url}/list-buckets",
            region="us-east-1",
            service="s3",
            access_key=self.access_key,
            secret_key=self.secret_key,
            payload_bytes=b"",
        )
        response = send_signed_request(signed_request)
        if response.status_code == 200:
            logger.info("Successfully retrieved bucket list.")
            try:
                root = ET.fromstring(response.content)
                return [
                    {'Name': b.findtext('Name'), 'Owner': b.findtext('Owner')}
                    for b in root.findall('Buckets')
                ]
            except ET.ParseError:
                logger.error("Failed to parse XML response from list-buckets.")
                return []
        else:
            logger.error(f"Error listing buckets: {response.status_code} {response.text}")
            return []

    def bucket_exists(self, bucket_name):
        return any(b['Name'] == bucket_name for b in self._list_buckets_raw())

    def get_bucket_owner(self, bucket_name):
        for bucket in self._list_buckets_raw():
            if bucket.get('Name') == bucket_name:
                return bucket.get('Owner')
        return None

    def change_bucket_owner(self, bucket_name, new_owner):
        url = f"{self.endpoint_url}/change-bucket-owner/?bucket={bucket_name}&owner={new_owner}"
        signed_request = sign_v4_request("PATCH", url, "us-east-1", "s3", self.access_key, self.secret_key, b"")
        response = send_signed_request(signed_request)
        if response.status_code >= 400:
            logger.error(f"Failed to change owner of bucket '{bucket_name}' to '{new_owner}': {response.status_code} {response.text}")
        else:
            logger.info(f"Successfully sent request to change owner of bucket '{bucket_name}' to '{new_owner}'.")

    def create_user(self, access_key, secret_key, role=None, user_id=None, group_id=None):
        if self.user_exists(access_key):
            logger.info(f"IAM user '{access_key}' already exists.")
            return
        root = ET.Element("Account")
        ET.SubElement(root, "Access").text = str(access_key)
        ET.SubElement(root, "Secret").text = str(secret_key)
        ET.SubElement(root, "Role").text = str(role) if role else "user"
        if user_id: ET.SubElement(root, "UserID").text = str(user_id)
        if group_id: ET.SubElement(root, "GroupID").text = str(group_id)
        
        xml_payload = ET.tostring(root)
        signed_request = sign_v4_request("PATCH", f"{self.endpoint_url}/create-user", "us-east-1", "s3", self.access_key, self.secret_key, xml_payload)
        response = send_signed_request(signed_request)
        if response.status_code >= 400:
            logger.error(f"Failed to create user '{access_key}': {response.status_code} {response.text}")
        else:
            logger.info(f"Successfully sent create user request for '{access_key}'.")

    def update_user(self, access_key, secret_key=None, user_id=None, group_id=None):
        root = ET.Element("MutableProps")
        if secret_key: ET.SubElement(root, "Secret").text = str(secret_key)
        if user_id: ET.SubElement(root, "UserID").text = str(user_id)
        if group_id: ET.SubElement(root, "GroupID").text = str(group_id)

        xml_payload = ET.tostring(root)
        url = f"{self.endpoint_url}/update-user?access={access_key}"
        signed_request = sign_v4_request("PATCH", url, "us-east-1", "s3", self.access_key, self.secret_key, xml_payload)
        response = send_signed_request(signed_request)
        if response.status_code >= 400:
            logger.error(f"Failed to update user '{access_key}': {response.status_code} {response.text}")
        else:
            logger.info(f"Successfully sent update request for user '{access_key}'.")

    def delete_user(self, access_key):
        url = f"{self.endpoint_url}/delete-user?access={access_key}"
        signed_request = sign_v4_request("PATCH", url, "us-east-1", "s3", self.access_key, self.secret_key, b"")
        response = send_signed_request(signed_request)
        if response.status_code >= 400:
            logger.error(f"Failed to delete user '{access_key}': {response.status_code} {response.text}")
        else:
            logger.info(f"Successfully deleted user '{access_key}'.")

    def _list_users_raw(self):
        signed_request = sign_v4_request("PATCH", f"{self.endpoint_url}/list-users", "us-east-1", "s3", self.access_key, self.secret_key, b"")
        response = send_signed_request(signed_request)
        if response.status_code == 200:
            try:
                root = ET.fromstring(response.content)
                return [
                    {'Access': a.findtext('Access')}
                    for a in root.findall('Accounts')
                ]
            except ET.ParseError:
                logger.error("Failed to parse XML response from list-users.")
                return []
        else:
            logger.error(f"Failed to list users: {response.status_code} {response.text}")
            return []

    def user_exists(self, access_key):
        return any(u['Access'] == access_key for u in self._list_users_raw())

File: pkg/backends/versitygw.go
package backends

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
	"github.com/aws/aws-sdk-go/service/s3"
	"k8s.io/klog/v2"
)

// VersityGW implements the Backend interface for VersityGW
type VersityGW struct {
	endpointURL string
	accessKey   string
	secretKey   string
	s3Client    *s3.S3
	httpClient  *http.Client
	signer      *v4.Signer
}

// NewVersityGW creates a new VersityGW backend
func NewVersityGW(config Config) *VersityGW {
	sess := session.Must(session.NewSession(&aws.Config{
		Endpoint:         aws.String(config.EndpointURL),
		Region:           aws.String("us-east-1"),
		Credentials:      credentials.NewStaticCredentials(config.AccessKey, config.SecretKey, ""),
		S3ForcePathStyle: aws.Bool(true),
	}))

	creds := credentials.NewStaticCredentials(config.AccessKey, config.SecretKey, "")

	return &VersityGW{
		endpointURL: config.EndpointURL,
		accessKey:   config.AccessKey,
		secretKey:   config.SecretKey,
		s3Client:    s3.New(sess),
		httpClient:  &http.Client{Timeout: 30 * time.Second},
		signer:      v4.NewSigner(creds),
	}
}

func (v *VersityGW) GetEndpointURL() string {
	return v.endpointURL
}

func (v *VersityGW) TestConnection(ctx context.Context) error {
	klog.Infof("Testing connection to VersityGW: %s", v.endpointURL)

	// Test listing buckets
	buckets, err := v.listBucketsRaw(ctx)
	if err != nil {
		return fmt.Errorf("failed to list buckets: %w", err)
	}
	klog.Infof("Successfully listed %d bucket(s)", len(buckets))

	// Test listing users
	users, err := v.listUsersRaw(ctx)
	if err != nil {
		return fmt.Errorf("failed to list users: %w", err)
	}
	klog.Infof("Successfully listed %d user(s)", len(users))

	return nil
}

func (v *VersityGW) CreateBucket(ctx context.Context, bucketName string, owner *string) error {
	exists, err := v.BucketExists(ctx, bucketName)
	if err != nil {
		return err
	}
	if exists {
		klog.Infof("Bucket %s already exists", bucketName)
		return nil
	}

	_, err = v.s3Client.CreateBucketWithContext(ctx, &s3.CreateBucketInput{
		Bucket: aws.String(bucketName),
	})
	if err != nil {
		return fmt.Errorf("failed to create bucket: %w", err)
	}

	klog.Infof("Created bucket: %s", bucketName)

	if owner != nil {
		return v.ChangeBucketOwner(ctx, bucketName, *owner)
	}

	return nil
}

func (v *VersityGW) DeleteBucket(ctx context.Context, bucketName string) error {
	_, err := v.s3Client.DeleteBucketWithContext(ctx, &s3.DeleteBucketInput{
		Bucket: aws.String(bucketName),
	})
	if err != nil {
		return fmt.Errorf("failed to delete bucket: %w", err)
	}

	klog.Infof("Deleted bucket: %s", bucketName)
	return nil
}

func (v *VersityGW) BucketExists(ctx context.Context, bucketName string) (bool, error) {
	_, err := v.s3Client.HeadBucketWithContext(ctx, &s3.HeadBucketInput{
		Bucket: aws.String(bucketName),
	})
	if err != nil {
		// Check if it's a "not found" error
		return false, nil
	}
	return true, nil
}

func (v *VersityGW) GetBucketOwner(ctx context.Context, bucketName string) (string, error) {
	url := fmt.Sprintf("%s/%s?acl", v.endpointURL, bucketName)
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return "", err
	}

	if err := v.signRequest(req, nil); err != nil {
		return "", err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to get bucket ACL: status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var acl struct {
		Owner struct {
			ID string `xml:"ID"`
		} `xml:"Owner"`
	}

	if err := xml.Unmarshal(body, &acl); err != nil {
		return "", err
	}

	return acl.Owner.ID, nil
}

func (v *VersityGW) ChangeBucketOwner(ctx context.Context, bucketName, newOwner string) error {
	url := fmt.Sprintf("%s/change-bucket-owner/?bucket=%s&owner=%s", v.endpointURL, bucketName, newOwner)

	req, err := http.NewRequestWithContext(ctx, "PATCH", url, nil)
	if err != nil {
		return err
	}

	if err := v.signRequest(req, nil); err != nil {
		return err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to change bucket owner: status %d, body: %s", resp.StatusCode, string(body))
	}

	klog.Infof("Changed owner of bucket %s to %s", bucketName, newOwner)
	return nil
}

func (v *VersityGW) CreateUser(ctx context.Context, accessKey, secretKey string, role *string, userID, groupID *int) error {
	exists, err := v.UserExists(ctx, accessKey)
	if err != nil {
		return err
	}
	if exists {
		klog.Infof("User %s already exists", accessKey)
		return nil
	}

	url := fmt.Sprintf("%s/create-user", v.endpointURL)

	userPayload := fmt.Sprintf(`<Account><Access>%s</Access><Secret>%s</Secret>`, accessKey, secretKey)
	if role != nil {
		userPayload += fmt.Sprintf(`<Role>%s</Role>`, *role)
	} else {
		userPayload += `<Role>user</Role>`
	}
	if userID != nil {
		userPayload += fmt.Sprintf(`<UserID>%d</UserID>`, *userID)
	}
	if groupID != nil {
		userPayload += fmt.Sprintf(`<GroupID>%d</GroupID>`, *groupID)
	}
	userPayload += `</Account>`

	payloadBytes := []byte(userPayload)

	req, err := http.NewRequestWithContext(ctx, "PATCH", url, bytes.NewReader(payloadBytes))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/xml")

	if err := v.signRequest(req, payloadBytes); err != nil {
		return err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to create user: status %d, body: %s", resp.StatusCode, string(body))
	}

	klog.Infof("Created user: %s", accessKey)
	return nil
}

func (v *VersityGW) DeleteUser(ctx context.Context, accessKey string) error {
	url := fmt.Sprintf("%s/delete-user?access=%s", v.endpointURL, accessKey)

	req, err := http.NewRequestWithContext(ctx, "PATCH", url, nil)
	if err != nil {
		return err
	}

	if err := v.signRequest(req, nil); err != nil {
		return err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to delete user: status %d, body: %s", resp.StatusCode, string(body))
	}

	klog.Infof("Deleted user: %s", accessKey)
	return nil
}

func (v *VersityGW) UpdateUser(ctx context.Context, accessKey string, secretKey *string, userID, groupID *int) error {
	url := fmt.Sprintf("%s/update-user?access=%s", v.endpointURL, accessKey)

	userPayload := `<MutableProps>`
	if secretKey != nil {
		userPayload += fmt.Sprintf(`<Secret>%s</Secret>`, *secretKey)
	}
	if userID != nil {
		userPayload += fmt.Sprintf(`<UserID>%d</UserID>`, *userID)
	}
	if groupID != nil {
		userPayload += fmt.Sprintf(`<GroupID>%d</GroupID>`, *groupID)
	}
	userPayload += `</MutableProps>`

	payloadBytes := []byte(userPayload)

	req, err := http.NewRequestWithContext(ctx, "PATCH", url, bytes.NewReader(payloadBytes))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/xml")

	if err := v.signRequest(req, payloadBytes); err != nil {
		return err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to update user: status %d, body: %s", resp.StatusCode, string(body))
	}

	klog.Infof("Updated user: %s", accessKey)
	return nil
}

func (v *VersityGW) UserExists(ctx context.Context, accessKey string) (bool, error) {
	users, err := v.listUsersRaw(ctx)
	if err != nil {
		return false, err
	}

	for _, user := range users {
		if user == accessKey {
			return true, nil
		}
	}

	return false, nil
}

func (v *VersityGW) listBucketsRaw(ctx context.Context) ([]string, error) {
	url := fmt.Sprintf("%s/list-buckets", v.endpointURL)
	req, err := http.NewRequestWithContext(ctx, "PATCH", url, nil)
	if err != nil {
		return nil, err
	}

	if err := v.signRequest(req, nil); err != nil {
		return nil, err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to list buckets: status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Buckets []struct {
			Name  string `xml:"Name"`
			Owner string `xml:"Owner"`
		} `xml:"Buckets"`
	}

	if err := xml.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	buckets := make([]string, len(result.Buckets))
	for i, b := range result.Buckets {
		buckets[i] = b.Name
	}

	return buckets, nil
}

func (v *VersityGW) listUsersRaw(ctx context.Context) ([]string, error) {
	url := fmt.Sprintf("%s/list-users", v.endpointURL)
	req, err := http.NewRequestWithContext(ctx, "PATCH", url, nil)
	if err != nil {
		return nil, err
	}

	if err := v.signRequest(req, nil); err != nil {
		return nil, err
	}

	resp, err := v.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to list users: status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Accounts []struct {
			Access string `xml:"Access"`
		} `xml:"Accounts"`
	}

	if err := xml.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	users := make([]string, len(result.Accounts))
	for i, u := range result.Accounts {
		users[i] = u.Access
	}

	return users, nil
}

func (v *VersityGW) signRequest(req *http.Request, body []byte) error {
	var bodyReader io.ReadSeeker
	if body != nil {
		bodyReader = bytes.NewReader(body)
		hash := sha256.Sum256(body)
		req.Header.Set("X-Amz-Content-Sha256", hex.EncodeToString(hash[:]))
	} else {
		bodyReader = bytes.NewReader([]byte{})
		req.Header.Set("X-Amz-Content-Sha256", "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855")
	}

	_, err := v.signer.Sign(req, bodyReader, "s3", "us-east-1", time.Now())
	return err
}
